README.md
# ideallical django-restframework testtools

## Requirements

* Python (3.5)

## Installation

Install using `pip`...

    pip install ii_drf_testtools

## Example

Let's take a look at a quick example of using ii_drf_testtools.

```python
import pytest

from ii_drf_testtools.utils import ListAPITest
from news.tests.factories import NewsItemFactory


@pytest.mark.django_db(transaction=True)
class TestNewsList(ListAPITest):
    api_url = reverse('rest_api:news:list', kwargs={'version': '1.0'})
    only_custom = False
    status_codes_anonymous = {'ALL': 403}
    status_codes_authenticated = {'ALL': 405, 'GET': 200}

    def setup_list_data(self):
        current_user = self.get_user()

        self.active_newsitem = NewsItemFactory(
            name='Active newsitem', is_active=True, user=current_user)
        self.active_newsitem2 = NewsItemFactory(
            name='Active newsitem2', is_active=True, user=current_user)
        self.inactive_newsitem = NewsItemFactory(
            name='Inactive newsitem', is_active=False, user=current_user)
        self.other_newsteam = NewsItemFactory(
            name='Other newsitem', is_active=True)

    def assert_list_response_authenticated(self, response):
        assert response.json() == [
            {
                'id': self.active_newsitem.id,
                'name': 'Active newsitem',
            },
            {
                'id': self.active_newsitem2.id,
                'name': 'Active newsitem2',
            }]
```

This test will do the following 10 tests:

    - Assert an anonymous GET request on /api/1.0/news/ equals status 403
    - Assert an anonymous DELETE request on /api/1.0/news/ equals status 403
    - Assert an anonymous POST request on /api/1.0/news/ equals status 403
    - Assert an anonymous PATCH request on /api/1.0/news/ equals status 403
    - Assert an anonymous PUT request on /api/1.0/news/ equals status 403

    - Assert an authenticated GET request on /api/1.0/news/ equals status 200
    - Assert an authenticated DELETE request on /api/1.0/news/ equals status 405
    - Assert an authenticated POST request on /api/1.0/news/ equals status 405
    - Assert an authenticated PATCH request on /api/1.0/news/ equals status 405
    - Assert an authenticated PUT request on /api/1.0/news/ equals status 405

Plus one additional test, namely:

    To test that a GET request on /api/1.0/news/ only shows active NewsItems
    that are created by the current loggedin user.


CHANGELOG.md
0.3 (2020-04-12)
----------------
 - Made package Django3 compatible and Python3 only

0.2 (2018-02-08)
----------------

 - Added CreateAPITest
 - Added GenericAPITest.only_custom to be able to skip default drf-testtools tests

0.1 (2017-09-29)
----------------

 - Project creation


ii_drf_testtools/utils.py
from ii_drf_testtools import mixins
from ii_drf_testtools.settings import testtools_settings
from rest_framework.test import APIClient

UserFactory = testtools_settings.USER_FACTORY


class GenericAPITest(object):

    # public variables
    api_url = None
    only_custom = False  # only test the custom methods (set to True to speed up testing)
    return_format = testtools_settings.RETURN_FORMAT
    status_codes_anonymous = testtools_settings.DEFAULT_STATUS_BASE_ANONYMOUS
    status_codes_authenticated = (
        testtools_settings.DEFAULT_STATUS_BASE_AUTHENTICATED)

    # internal variables
    _anonymous_client = None
    _user = None

    def get_api_url(self, **kwargs):
        if self.api_url is None:  # pragma: no cover
            raise NotImplementedError('Please provide url')
        return self.api_url

    def get_user(self):
        """
        Returns user object
        """
        if self._user is None:
            self._user = UserFactory()
        return self._user

    def get_expected_status_code(self, status_dict, method, status_code=None, **kwargs):
        if status_code:
            return status_code
        try:
            return status_dict[method]
        except KeyError:
            try:
                return status_dict['ALL']
            except KeyError:
                raise NotImplementedError('Please implement status in dict {} '
                                          'using method {}'.format(
                                              status_dict, method))

    def assert_status_code(self, response, status_dict, method,
                           status_code=None, **kwargs):

        expected_status_code = self.get_expected_status_code(
            status_dict, method, status_code)

        assert response.status_code == expected_status_code, (
            'A {method} request on {api_url} returned {status_code} but should'
            ' be {expected_status_code}.'.format(
                method=method,
                api_url=self.get_api_url(**kwargs),
                status_code=response.status_code,
                expected_status_code=expected_status_code))

    def get_anonymous_client(self):
        """
        Returns non authenticated client
        """
        if self._anonymous_client is None:  # pragma: no cover
            self._anonymous_client = APIClient()
        return self._anonymous_client

    def get_authenticated_client(self, user=None):
        """
        Returns authenticated client with a normal user
        """
        if user is None:
            user = self.get_user()

        _authenticated_client = APIClient()
        _authenticated_client.force_authenticate(user=user)
        return _authenticated_client

    def test_status_on_anonymous_get(self, status_code=None, **kwargs):
        """
        Test the correct response status_code for a GET request
        that is not authenticated
        """
        if not self.only_custom:
            response = self.get_anonymous_client().get(
                path=self.get_api_url(**kwargs), format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_anonymous, 'GET', status_code,
                **kwargs)

    def test_status_on_anonymous_delete(self, status_code=None, **kwargs):
        """
        Test the correct response status_code for a DELETE request
        that is not authenticated
        """
        if not self.only_custom:
            response = self.get_anonymous_client().delete(
                path=self.get_api_url(**kwargs), format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_anonymous, 'DELETE', status_code,
                **kwargs)

    def test_status_on_anonymous_post(self, status_code=None, **kwargs):
        """
        Test the correct response status_code for a POST request
        that is not authenticated
        """
        if not self.only_custom:
            response = self.get_anonymous_client().post(
                path=self.get_api_url(**kwargs), format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_anonymous, 'POST', status_code,
                **kwargs)

    def test_status_on_anonymous_patch(self, status_code=None, **kwargs):
        """
        Test the correct response status_code for a PATCH request
        that is not authenticated
        """
        if not self.only_custom:
            response = self.get_anonymous_client().patch(
                path=self.get_api_url(**kwargs), format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_anonymous, 'PATCH', status_code,
                **kwargs)

    def test_status_on_anonymous_put(self, status_code=None, **kwargs):
        """
        Test the correct response status_code for a PUT request
        that is not authenticated
        """
        if not self.only_custom:
            response = self.get_anonymous_client().put(
                path=self.get_api_url(**kwargs), format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_anonymous, 'PUT', status_code,
                **kwargs)

    def test_status_on_authenticated_get(self, status_code=None, **kwargs):
        """
        Test the correct response status_code for a GET request
        that is authenticated
        """
        if not self.only_custom:
            response = self.get_authenticated_client().get(
                path=self.get_api_url(**kwargs), format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_authenticated, 'GET', status_code,
                **kwargs)

    def test_status_on_authenticated_delete(self, status_code=None, **kwargs):
        """
        Test the correct response status_code for a DELETE request
        that is authenticated
        """
        if not self.only_custom:
            response = self.get_authenticated_client().delete(
                path=self.get_api_url(**kwargs), format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_authenticated, 'DELETE',
                status_code, **kwargs)

    def test_status_on_authenticated_post(self, status_code=None,
                                          data=None, **kwargs):
        """
        Test the correct response status_code for a POST request
        that is authenticated
        """
        if not self.only_custom:
            response = self.get_authenticated_client().post(
                path=self.get_api_url(**kwargs),
                data=data, format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_authenticated, 'POST', status_code,
                **kwargs)

    def test_status_on_authenticated_patch(self,
                                           status_code=None, data=None,
                                           **kwargs):
        """
        Test the correct response status_code for a PATCH request
        that is authenticated
        """
        if not self.only_custom:
            response = self.get_authenticated_client().patch(
                path=self.get_api_url(**kwargs), data=data,
                format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_authenticated, 'PATCH',
                status_code, **kwargs)

    def test_status_on_authenticated_put(self,
                                         status_code=None, data=None,
                                         **kwargs):
        """Test the correct response status_code for a PUT request
           that is authenticated.
        """
        if not self.only_custom:
            response = self.get_authenticated_client().put(
                path=self.get_api_url(**kwargs), data=data,
                format=self.return_format)
            self.assert_status_code(
                response, self.status_codes_authenticated, 'PUT', status_code,
                **kwargs)


class ListAPITest(mixins.ListModelMixin, GenericAPITest):
    status_codes_anonymous = testtools_settings.DEFAULT_STATUS_LIST_ANONYMOUS
    status_codes_authenticated = (
        testtools_settings.DEFAULT_STATUS_LIST_AUTHENTICATED)


class CreateAPITest(GenericAPITest):
    status_codes_anonymous = testtools_settings.DEFAULT_STATUS_CREATE_ANONYMOUS
    status_codes_authenticated = (
        testtools_settings.DEFAULT_STATUS_CREATE_AUTHENTICATED)


class DeleteAPITest(GenericAPITest):
    status_codes_anonymous = testtools_settings.DEFAULT_STATUS_DELETE_ANONYMOUS
    status_codes_authenticated = (
        testtools_settings.DEFAULT_STATUS_DELETE_AUTHENTICATED)


class RetrieveAPITest(GenericAPITest):
    status_codes_anonymous = (
        testtools_settings.DEFAULT_STATUS_RETRIEVE_ANONYMOUS)
    status_codes_authenticated = (
        testtools_settings.DEFAULT_STATUS_RETRIEVE_AUTHENTICATED)


class UpdateAPITest(GenericAPITest):
    status_codes_anonymous = testtools_settings.DEFAULT_STATUS_UPDATE_ANONYMOUS
    status_codes_authenticated = (
        testtools_settings.DEFAULT_STATUS_UPDATE_AUTHENTICATED)


ii_drf_testtools/mixins.py
class ListModelMixin(object):

    def setup_list_data(self):
        raise NotImplementedError('Please implement setup_list_data')

    def assert_list_response_anonymous(self):
        raise NotImplementedError(
            'Please implement assert_list_response_anonymous')

    def assert_list_response_authenticated(self):
        raise NotImplementedError(
            'Please implement assert_list_response_authenticated')

    def test_data_on_anonymous_get(self, **kwargs):
        if not self.only_custom:
            expected_status_code = self.get_expected_status_code(
                self.status_codes_anonymous, 'GET', **kwargs)

            if expected_status_code == 200:
                self.setup_list_data()
                response = self.get_anonymous_client().get(
                    path=self.get_api_url(**kwargs), format=self.return_format)
                self.assert_list_response_anonymous(response)

    def test_data_on_authenticated_get(self, **kwargs):
        if not self.only_custom:
            expected_status_code = self.get_expected_status_code(
                self.status_codes_authenticated, 'GET', **kwargs)

            if expected_status_code == 200:
                self.setup_list_data()
                response = self.get_authenticated_client().get(
                    path=self.get_api_url(**kwargs), format=self.return_format)
                self.assert_list_response_authenticated(response)


ii_drf_testtools/settings.py
"""
Settings for ideallical drf testtools are all namespaced in the
II_DRF_TESTTOOLS setting.
For example your project's `settings.py` file might look like this:

II_DRF_TESTTOOLS = {
    'RETURN_FORMAT': 'json',
    'USER_FACTORY': 'account.tests.factories.UserFactory',
    # base
    'DEFAULT_STATUS_BASE_ANONYMOUS': {'ALL': 403, },
    'DEFAULT_STATUS_BASE_AUTHENTICATED': {'ALL': 405, },
    # list
    'DEFAULT_STATUS_LIST_ANONYMOUS': {'ALL': 403, },
    'DEFAULT_STATUS_LIST_AUTHENTICATED': {'ALL': 405, 'GET': 200},
    # create
    'DEFAULT_STATUS_CREATE_ANONYMOUS': {'ALL': 403, },
    'DEFAULT_STATUS_CREATE_AUTHENTICATED': {'ALL': 405, 'POST': 400},
    # delete
    'DEFAULT_STATUS_DELETE_ANONYMOUS': {'ALL': 403, },
    'DEFAULT_STATUS_DELETE_AUTHENTICATED': {'ALL': 405, 'DELETE': 404},
    # get
    'DEFAULT_STATUS_RETRIEVE_ANONYMOUS': {'ALL': 403, },
    'DEFAULT_STATUS_RETRIEVE_AUTHENTICATED': {'ALL': 405, 'GET': 200},
    # update
    'DEFAULT_STATUS_UPDATE_ANONYMOUS': {'ALL': 403, },
    'DEFAULT_STATUS_UPDATE_AUTHENTICATED': {'ALL': 405, 'PUT': 400}
}


This module provides the `testtool_setting` object, that is used to access
ideallical DRF testtools settings, checking for user settings first, then
falling back to the defaults.
"""
from ii_django_package_settings.settings import PackageSettings


class TestToolsSettings(PackageSettings):
    NAME = 'II_DRF_TESTTOOLS'
    DOC = 'https://github.com/ideallical/ii_drf_testtools/'
    DEFAULTS = {
        'RETURN_FORMAT': 'json',
        'USER_FACTORY': 'accounts.tests.factories.UserFactory',
        # base
        'DEFAULT_STATUS_BASE_ANONYMOUS': {'ALL': 403, },
        'DEFAULT_STATUS_BASE_AUTHENTICATED': {'ALL': 405, },
        # list
        'DEFAULT_STATUS_LIST_ANONYMOUS': {'ALL': 403, },
        'DEFAULT_STATUS_LIST_AUTHENTICATED': {'ALL': 405, 'GET': 200},
        # create
        'DEFAULT_STATUS_CREATE_ANONYMOUS': {'ALL': 403, },
        'DEFAULT_STATUS_CREATE_AUTHENTICATED': {'ALL': 405, 'POST': 400},
        # delete
        'DEFAULT_STATUS_DELETE_ANONYMOUS': {'ALL': 403, },
        'DEFAULT_STATUS_DELETE_AUTHENTICATED': {'ALL': 405, 'DELETE': 404},
        # get
        'DEFAULT_STATUS_RETRIEVE_ANONYMOUS': {'ALL': 403, },
        'DEFAULT_STATUS_RETRIEVE_AUTHENTICATED': {'ALL': 405, 'GET': 200},
        # update
        'DEFAULT_STATUS_UPDATE_ANONYMOUS': {'ALL': 403, },
        'DEFAULT_STATUS_UPDATE_AUTHENTICATED': {'ALL': 405, 'PUT': 400}
    }
    IMPORT_STRINGS = ('USER_FACTORY', )


testtools_settings = TestToolsSettings(None)
